vision/diffusion_mnist/diffusion_mnist.jl

```diff
@@ -0,0 +1,330 @@
+# Score-Based Generative Modeling
+#
+# Score-Based Generative Modeling through Stochastic Differential Equations.
+# Yang Song, Jascha Sohl-Dickstein, Diederik P. Kingma, 
+# Abhishek Kumar, Stefano Ermon, and Ben Poole.
+# https://arxiv.org/pdf/2011.13456.pdf
+
+using MLDatasets
+using Flux
+using Flux: @functor, chunk, params
+using Flux.Data: DataLoader
+using Parameters: @with_kw
+using BSON
+using CUDA
+using Images
+using Logging: with_logger
+using ProgressMeter: Progress, next!
+using TensorBoardLogger: TBLogger, tb_overwrite
+using Random
+
+"""
+Projection of Gaussian Noise onto a time vector.
+
+# Notes
+This layer will help embed our random times onto the frequency domain. \n
+W is not trainable and is sampled once upon construction - see assertions below.
+
+# References
+paper-  https://arxiv.org/abs/2006.10739 \n
+layers- https://fluxml.ai/Flux.jl/stable/models/basics/#Building-Layers
+"""
+function GaussianFourierProjection(embed_dim, scale)
+    # Instantiate W once
+    W = randn(Float32, embed_dim ÷ 2) .* scale
+    # Return a function that always references the same W
+    function GaussFourierProject(t)
+        t_proj = t' .* W * Float32(2π)
+        [sin.(t_proj); cos.(t_proj)]
+    end
+end
```

Comment:
If `W` is learnable, this should be a struct type. Ref. https://fluxml.ai/Flux.jl/stable/models/advanced/.

Reply:
Yes, `W` is purposely not learnable.